### PR TITLE
[5.0] PSR-2 AND PSR-5

### PR DESCRIPTION
**Pull request by GrahamCampbell**

### CODE REVIEW

**Comment by Anahkiasen** [comment]
:+1:

**Comment by RomainLanz** [comment]
:+1: 

What do you think about keep the namespace on the same line that the `<?php` tag?

**Comment by GrahamCampbell** [comment]
PSR-2 doesn't allow it unfortunately.

**Comment by GrahamCampbell** [comment]
It says it must be on a new line, and common convention is to leave another line between that opening tag and the namespace, so I did that.

**Comment by RomainLanz** [comment]
I never see that is not possible for the namespace. The only one rule for the position is : `When present, there MUST be one blank line after the namespace declaration.`

````
When present, there MUST be one blank line after the namespace declaration.

When present, all use declarations MUST go after the namespace declaration.

There MUST be one use keyword per declaration.

There MUST be one blank line after the use block.
````

**Comment by Anahkiasen** [comment]
It's not in PSR2 but to be fair apart from in Laravel I've never seen it on the same line and since the point here is to go towards unification with other frameworks, I'd say putting it on a new line would be best no?

**Comment by GrahamCampbell** [comment]
Agreed.

**Comment by JoostK** [comment]
Why does PSR-2 have to be so ugly :cry:

**Comment by GrahamCampbell** [comment]
PSR-2 is pretty :) I've got used to it.

**Comment by RomainLanz** [comment]
PSR-2 is not so ugly.

Ok, no problem for the namespace, I just think that it's better to have it on the same line.

**Comment by hannesvdvreken** [comment]
:+1:

**Comment by KennedyTedesco** [comment]
Beautiful!

**Comment by shin1x1** [comment]
:+1:

**Comment by tomschlick** [comment]
:+1:

File: src/Illuminate/Auth/DatabaseUserProvider.php [hunk 1]
```diff
@@ -1,146 +1,151 @@
...
+     * @param string $token
+     *
+     * @return \Illuminate\Contracts\Auth\Authenticatable|null
+     */
+    public function retrieveByToken($identifier, $token)
+    {
+        $user = $this->conn->table($this->table)
                                 ->where('id', $identifier)
                                 ->where('remember_token', $token)
                                 ->first();
```
Comment (ibrasho) [inline src/Illuminate/Auth/DatabaseUserProvider.php]: Alignment?

File: src/Illuminate/Auth/DatabaseUserProvider.php [hunk 2]
```diff
@@ -1,146 +1,151 @@
...
+     * Update the "remember me" token for the given user in storage.
+     *
+     * @param \Illuminate\Contracts\Auth\Authenticatable $user
+     * @param string                                     $token
+     *
+     * @return void
+     */
+    public function updateRememberToken(UserContract $user, $token)
+    {
+        $this->conn->table($this->table)
```
Comment (ibrasho) [inline src/Illuminate/Auth/DatabaseUserProvider.php]: Alignment?

File: src/Illuminate/Auth/Guard.php [hunk 3]
```diff
@@ -9,755 +11,780 @@
...
+     * Attempt to authenticate using basic authentication.
+     *
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     * @param string                                    $field
+     *
+     * @return bool
+     */
+    protected function attemptBasic(Request $request, $field)
+    {
+        if (! $request->getUser()) {
```
Comment (ibrasho) [inline src/Illuminate/Auth/Guard.php]: Should not this be `if ( ! $request->getUser()) {`?

Comment (RomainLanz) [inline src/Illuminate/Auth/Guard.php]: The exclamation point is not reglemented on PSR-2, I think it's more readable to keep a space before and after it too.

What do you think @GrahamCampbell

Comment (crynobone) [inline src/Illuminate/Auth/Guard.php]: It's part of PSR-2

> Opening parentheses for control structures MUST NOT have a space after them, and closing parentheses for control structures MUST NOT have a space before.

Comment (crynobone) [inline src/Illuminate/Auth/Guard.php]:
```php
if ( $foo ) { // wrong
```

```php
if ($foo) { // right
```

```php
if (! $foo) { // is not wrong :P
```

File: src/Illuminate/Auth/DatabaseUserProvider.php (hunk 2 above)
Comment (GrahamCampbell) [inline src/Illuminate/Auth/DatabaseUserProvider.php]: the alignment is correct here

File: src/Illuminate/Auth/Guard.php (hunk 3 above)
Comment (GrahamCampbell) [inline src/Illuminate/Auth/Guard.php]: Yeh, this is fine as it is.

File: src/Illuminate/Auth/DatabaseUserProvider.php (hunk 1 above)
Comment (GrahamCampbell) [inline src/Illuminate/Auth/DatabaseUserProvider.php]: I have no automated way to fix such cases. Also, I don't think PSR-2 says anything about this.

**Comment by anhskohbo** [comment]
:+1:

File: src/Illuminate/Auth/DatabaseUserProvider.php (hunk 1 above)
Comment (ibrasho) [inline src/Illuminate/Auth/DatabaseUserProvider.php]: I know, but I think we want to keep the Laravel-ish feel. :-)

**Comment by taylorotwell** [comment]
Rebase? :)

**Comment by GrahamCampbell** [comment]
Will do when I find a min.

**Comment by hannesvdvreken** [comment]
Cool! :+1:

**Comment by jwpage** [comment]
Just curious, will there be an accompanying PR for laravel/laravel so that the code there is PSR-2+5 as well, or is this only going to be a standard for laravel/framework?

**Comment by xiaobeicn** [comment]
:+1:

**Comment by alioygur** [comment]
:+1:

**Comment by alioygur** [comment]
@GrahamCampbell it should be on laravel/laravel

File: src/Illuminate/Auth/Guard.php (hunk 3 above)
Comment (gaomd) [inline src/Illuminate/Auth/Guard.php]: `if ( ! $foo) { // seems more readable`

**Comment by GrahamCampbell** [comment]
Yeh, I will do it there too. I'll do it once Taylor's totally happy with the default app.

**Comment by GrahamCampbell** [comment]
And indeed on the other laravel repos.

**Comment by GrahamCampbell** [comment]
NB, I'll be rebasing this once 4.2.17 is tagged, and merged with the master branch.

**Comment by zombopanda** [comment]
:thumbsup:

**Comment by GrahamCampbell** [comment]
I'll resend this at a later point when we're truly ready to do this. The merging of this pull will mean that Laravel 4.2 is no-longer maintained for anything but more important bug fixes, so we'll be doing this close to the 5.0.0 release.

**Comment by JosephSilber** [comment]
@GrahamCampbell at that cutoff point, shouldn't we also convert all arrays to the short syntax?

**Comment by ibrasho** [comment]
^ I agree.

**Comment by GrahamCampbell** [comment]
> @GrahamCampbell at that cutoff point, shouldn't we also convert all arrays to the short syntax?

I already did that as part of this change, and yeh, I will be doing that in the replacement too.

**Comment by GrahamCampbell** [comment]
NB, only on 5.0.

**Comment by HipsterJazzbo** [comment]
Is this going to happen for release? Can't stand the suspense!

**Comment by overtrue** [comment]
PSR-2 is pretty :+1:

**Comment by daids** [comment]
:+1:

**Comment by GrahamCampbell** [comment]
This is not ever going to happen, sorry.

**Comment by overtrue** [comment]
@GrahamCampbell What a sad ending... :cry:

File: src/Illuminate/Auth/Guard.php (hunk 3 above)
Comment (gregrobson) [inline src/Illuminate/Auth/Guard.php]: Personally I have found that removing the negation from the if statement is the best solution as it improves readability.
```php
if (! $request->getUser()) {
// vs
if ($request->noUserSpecified()) {
```
I haven't used the above method in my own code, but hopefully it demonstrates the idea. as another example it's far easier to say *is empty* than *does not have more than 0 items*

**Comment by GrahamCampbell** [comment]
I'm a bit gutted too. Taylor does have a point that this whole cs thing is a total waste of time.

The big issue for me is not that laravel/framework's code isn't PSR-2, but all the generated files arn;t PSR-2, which means I have to change them every time. :(

**Comment by Anahkiasen** [comment]
> Taylor does have a point that this whole cs thing is a total waste of time

Uh I don't think that's a good argument, I'd rather have it be a waste of time _once_ and then being able to let a tool fix them every time

**Comment by MaartenStaa** [comment]
I have to agree with Maxime. It's quite a bit of work up front, but once you're there, there's a lot of nice tools for enforcing PSR-2.

It personally took me a while to get used to the standard, but now I think it's a prettier CS than what Laravel uses. Plus, since so many packages/frameworks use it, that's really a big plus IMO.

**Comment by crynobone** [comment]
> I'm a bit gutted too. Taylor does have a point that this whole cs thing is a total waste of time.

It would be useful if Laravel 4.2 still support minor bugfix (instead of security fix). Expecting everyone to jump to 5 straight away when they have production code when what not available right now in 4.2 is some minor bugs IMHO is not acceptable. The idea right now is Laravel 5 going to be available today (update now as Laravel 4 is EOL unless there some security issue).

p/s: I am under the assumption that 4 become security only due to moving to PSR-2.

**Comment by barryvdh** [comment]
It's a total waste of time having to explain the differences of the Laravel CS to everyone, commenting on PR's, not having to validate with proper tools, no default settings in most IDE's etc.
Regular contributors know the difference, but it's annoying to keep on switching. External contributors usually just assume PSR-2..

If Taylor doesn't care, but a lot of contributors do, why not listen to them?

**Comment by GrahamCampbell** [comment]
I think the point he's making is it doesn't really matter if the cs of contributions is incorrect.

He's made up his mind already anyway. I want everything to go PSR-2, but he doesn't.

**Comment by barryvdh** [comment]
> I think the point he's making is it doesn't really matter if the cs of contributions is incorrect.

Wait, are you saying we can just submit PR's in any CS we want? So just submitting stuff in PSR-2 is okay too? ;)

**Comment by GrahamCampbell** [comment]
No, it's not ok. Hmmm. This is a mess if you ask me. :/

**Comment by Garbee** [comment]
imo, if a PR is sent with improper CS it would be just as quick to pull the PR in locally, fix the CS and push it up then it would be commenting on every little damn thing. However, that would end up being a huge time resource with all the little issues found then it would be to just go PSR-2 and let tools handle it.

Everyone is correct, this does a major hurt exactly once, aside from needing to backport fixes since it ends up making more work to do that between 5.x and 4.x. But once that is done everything is more seamless for the majority of developers using the platform.

Moving to PSR-2 could also lead to more contributions back since it is less friction to make sure you are coding to the Laravel standard instead of PSR-2 (or any other code style that has a wide array of tooling available.)

However, it looks like 5.0 has gone stable within the past few hours with all the site updates and all. So this shipped has sailed for this major. Maybe with 6.0...

**Comment by Anahkiasen** [comment]
Honestly considering how different the L5 and L4 codebases are now I don't think CS are would even be the main reasons backports are hard

**Comment by GrahamCampbell** [comment]
If Taylor did change, he'd only do it on the framework repo, and not the other repos, which I also think is very strange. It's to late now anyway.

**Pull request closed without merging**